API 테스트 코드를 작성하라

tlatmsrud opened the pull request:
모든 코드에 대한 테스트 코드를 작성하라.

tlatmsrud committed:
API 테스트 코드를 작성하라

모든 코드에 대한 테스트 코드를 작성하라.

tlatmsrud commented:
멘토님, Test 코드 작성이 처음이라 아주 원초적으로 궁금한게 있습니다.
mockBean 을 통해 가상의 객체를 만들고 그 객체에 대한 response 값을 given으로 정해주고 있는데요, 이게 결국 API에 대한 응답을 실제 코드가 아닌 개발자가 정하는 구조이기에 개발자의 생각(?)에 따라 테스트 결과가 달라질 것이고, 이는 신뢰성이 떨어지는 테스트가 아닐까 라는 의구심이 강하게 듭니다.
TDD 가 익숙하지 않아 드는 생각이라는 확신도 강하게 듭니다만 이 의구심을 떨쳐버릴 수 없습니다 ㅠ
멘토님의 생각을 조금만 덜어주실 수 있으실까요 ? ㅠ

hannut91 commented:
Mocking을 하다보니 당연히 테스트가 통과할 것이고 이런 테스트가 가치가 없어 보일 수 있죠.

Mocking은 테스트를 굉장히 단순하게 만들어주고 테스트 대상을 명확하게 해줍니다. 그리고 구현에 관한 결정도 미루게 해줍니다. 예를들어 로그인을 만들고 있는데, 인증을 처리해줄 무언가가 필요합니다. 이런 구현을 어떻게 할지 모르겠어서 Mocking으로 기대하는 바로만 동작하도록 하여 로그인 기능을 만들고, 인증을 처리해줄 무언가를 실제로 구현할 수 있죠.

그런데 걱정하신대로 실질적으로 무언가가 조립되었을 때 올바르게 동작하는지는 테스트하기 어려워집니다. 각각은 잘 동작하지만 조립했을 때 잘 동작하지 않을 수 있죠.

![483bcd0cd4c31fca7d3f66c3119f099c51d877ec](https://user-images.githubusercontent.com/14071105/220558095-429b77e7-5155-4a8a-a226-5f57bfb08c2f.gif)

그래서 통합테스트를 하지 않으면 위와 같은 일이 벌어집니다. 유닛 테스트 -> 통합테스트 -> E2E 테스트로 갈수록 사용자에게 더 가치가 높습니다.

그래서 저는 둘다 작성합니다. Unit 테스트도 작성하고 통합테스트도 작성합니다. TDD를 하다보면 Mock없이 힘들 때가 많습니다. 그리고 나중에 통합테스트로 전환하기도 합니다. 이런 내용에 대해서는 다음과 같은 책에서 아주 잘 다루고 있습니다.

https://product.kyobobook.co.kr/detail/S000001032969

tlatmsrud commented:
말씀 감사드립니다! 추가로 여쭤볼게 있는데요, TaskControllerWebTest의 setUp 메서드에 mockBean에 대한 응답값을 설정하는 코드를 모두 몰아넣었는데요. 이렇게 한 메서드에 given을 몰아넣는게 과연 괜찮은 방법일까라는 궁금증이 듭니다. 제안해주실 부분이 있으실까요>??

hannut91 commented:
> TaskControllerWebTest의 setUp 메서드에 mockBean에 대한 응답값을 설정하는 코드를 모두 몰아넣었는데요. 이렇게 한 메서드에 given을 몰아넣는게 과연 괜찮은 방법일까라는 궁금증이 듭니다. 제안해주실 부분이 있으실까요>??

간단한 테스트의 경우 몰아넣으면 테스트 부분이 단순해질수가 있죠. 그리고 주어진 상황이 복잡하지 않아서 몰아 넣어도 충분한 경우가 있습니다. 하지만 주어진 상황이 많아지면 감당하기가 어려워지고 알아보기도 어려워지겠죠. 저는 주어진 상황이 달라진 것을 명확하게 보여주고 싶은 경우 명확하게 표현합니다. 예를들어서 달라지는 상황을 명확하게 표현하려고 할 수도 있습니다.

```java
@Nested
class With_authenticated_user {
    @BeforeEach
    void setUp() {
        given(authService.login).willReturn(authentication);
    }

    void someTest() {
       // ... 생략
    }
}

@Nested
class With_unauthenticated_user {
    @BeforeEach
    void setUp() {
        given(authService.login).thenThrow(UnauthenticatedException.class);
    }

    void someTest() {
       // ... 생략
    }
}
```

tlatmsrud committed:
@Nested를 사용하여 테스트를 계층적으로 분리하라

1. @Nested 어노테이션을 사용하여 계층적으로 분리합니다. 이로써
   mockBean에 대한 given절을 기능별로 분리하게 되었습니다.
2. @DisplayName 어노테이션을 사용하여 테스트 내용을 명확히합니다.

tlatmsrud commented:
@Nested를 통해 TaskControllerWebTest 코드를 분리해보았습니다. 그런데 예상못한 에러가 발생하는 곳이 있더라구요. 디버깅을 해보니 TaskControllerWebTest의 114라인 setUp메서드 실행 시 given(taskService.getTask(INVALID_ID)).willThrow(new TaskNotFoundException(INVALID_ID)); 구문에서 TaskNotFoundException 이 발생해버립니다...

![image](https://user-images.githubusercontent.com/9374562/220972085-6c7f9600-df82-484d-8376-112fc54026e7.png)

참고로 TaskDetailTest클래스에 대한 테스트 실행 시 발생하며, 메서드 단위로 각각 테스트를 진행할 경우 발생하지 않았습니다.

이에 대한 원인은 발견하지 못하였습니다만, @beforeEach로 인해 해당 구문이 두번째 실행될때 에러가 발생하므로 구문을detailWithInvalidId 메서드 안으로 옮겨주니 오류는 발생하지 않았습니다. 혹시 원인을 알려주실 수 있으신가요?

hannut91 commented:
지금 사례에 맞는지는 확실하진 않지만, mock객체가 초기화가 되지 않아서 그런걸로 기억합니다. 이전 모킹에서 예외를 던지도록 했고, 그게 남아있어서 예외를 던져서 안되는 걸로 기억해요

https://github.com/CodeSoom/spring-week5-assignment-1/pull/81#issuecomment-1341906184

위 내용을 참고하시면 좋을 것 같아요

See also
  - https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#verification

tlatmsrud commented:
삭제된 댓글입니다!
